adapters/medianet/medianet.go

```diff
@@ -0,0 +1,139 @@
+	var errs []error
+
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: "No impression in the request",
+		}}
+	}
+
+	if err := preprocess(&request.Imp[0]); err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	reqJson, err := json.Marshal(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqJson,
+		Headers: headers,
+	}}, errs
+}
+
+func (a *MedianetAdapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	var errs []error
+
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb2.BidResponse
+
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponse()
+
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bidType, err := getMediaTypeForImp(sb.Bid[i].ImpID, internalRequest.Imp)
+			if err != nil {
+				errs = append(errs, err)
+			} else {
+				b := &adapters.TypedBid{
+					Bid:     &sb.Bid[i],
+					BidType: bidType,
+				}
+				bidResponse.Bids = append(bidResponse.Bids, b)
+			}
+		}
+	}
+	return bidResponse, errs
+}
+
+// Builder builds a new instance of the Medianet adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	return &MedianetAdapter{
+		endpoint: config.Endpoint,
+	}, nil
+}
+
+func preprocess(imp *openrtb2.Imp) error {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var medianetExt openrtb_ext.ExtImpMedianet
+	if err := json.Unmarshal(bidderExt.Bidder, &medianetExt); err != nil {
+		return &errortypes.BadInput{Message: "bad Medianet bidder ext"}
+	}
+
+	if len(medianetExt.Cid) == 0 && len(medianetExt.Crid) == 0 {
+		return &errortypes.BadInput{Message: "'cid' and 'crid' are required attribute for Medianet's bidder ext"}
+	} else if len(medianetExt.Cid) == 0 {
+		return &errortypes.BadInput{Message: "'cid' is required attribute for Medianet's bidder ext"}
+	} else if len(medianetExt.Crid) == 0 {
```

Comment:
Yes, Thanks for pointing this out, I will add it in bidder-params but for custom error message (specific to medianet) this code will still be required.
Elaboration:
For test case `medianettest/supplemental/invaid-req-missing-cid-crid-error.json`
In the above test case I will get error `"'cid' and 'crid' are required attribute for Medianet's bidder ext"` but if I remove those checks from code and add it in bidder-param json file I will get `MakeRequests had wrong error count. Expected 1, got 0 ([])` which is not quite specific

Reply:
Ok. If the custom error messages are required, lets leave the checks in lines 110 to 116 and not add the `"minLength"` entries to `static/bidder-params/medianet.json` because those entries would make the true banches of said `if` statements unreachable.

Thank you for clarifying this point.

Reply:
No error is expected in this case. Request.imp.ext validation with bidder param schema happens in Action code: action.go -> func validateImpExt -> deps.paramsValidator.Validate. In unit tests, bidder code (bidder.MakeRequests) executes directly from unit test itself. This is why json schema validation will not work. 
There is a params_test.go to do this. 

For real request this code will never be executed and any of those messages will ever be returned because of json schema validation happens first.

Reply:
Thanks